cranelift/frontend/src/ssa.rs

```diff
@@ -307,11 +312,11 @@ impl SSABuilder {
                 return true;
             }
 
-            if visited.contains(&current) {
+            if self.visited.contains(&current) {
                 return false;
             }
-            visited.insert(current);
             current = predecessors[0].block;
+            self.visited.insert(current);
```

Comment:
Huh, moving the `insert` after changing `current` doesn't seem right. Did you mean to do that?

A pre-existing nit: `insert` returns true if `contains` would have returned false. So assuming you don't want to change the order here those two calls can be collapsed into `if !self.visited.insert(current) { ...`

Reply:
> Huh, moving the `insert` after changing `current` doesn't seem right. Did you mean to do that?

Have to to appease the borrow checker, since `predecessors` has a shared borrow of `self`.

Reply:
Ah but I did introduce a bug, I see now what you mean, will fix in a sec.

Reply:
The simplest fix is to move the `visited` check before the `predecessors.len()` checks: if we've visited this block before, it already passed the `predecessors.len()` checks.